File: cpp/p74.cpp
#include <cassert>
#include <vector>

using namespace std;

class Solution {
public:
    bool searchMatrix(vector<vector<int>> &matrix, int target) {
        int m = matrix.size(), n = matrix[0].size();
        int i = m - 1, j = 0;
        while (i >= 0 && j < n) {
            if (matrix[i][j] == target) {
                return true;
            } else if (matrix[i][j] < target) {
                j += 1;
            } else {
                i -= 1;
            }
        }
        return false;
    }
};

int main() {
    vector<vector<int>> matrix = {{1, 4, 7, 11, 15},
                                  {2, 5, 8, 12, 19},
                                  {3, 6, 9, 16, 22},
                                  {10, 13, 14, 17, 24},
                                  {18, 21, 23, 26, 30}};
    assert(Solution().searchMatrix(matrix, 5));
    assert(!Solution().searchMatrix(matrix, 20));
    return 0;
}

File: rust/src/problems/p74.rs
use std::cmp::Ordering;

pub struct Solution;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let m = matrix.len();
        let n = matrix[0].len();
        let get = |idx: usize| matrix[idx / n][idx % n];

        let mut left = 0i32;
        let mut right = (m * n - 1) as i32;
        while left >= 0 && right < (m * n) as i32 && left <= right {
            let c = left + (right - left) / 2;
            match get(c as usize).cmp(&target) {
                Ordering::Greater => right = c - 1,
                Ordering::Less => left = c + 1,
                Ordering::Equal => return true,
            }
        }
        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let matrix = vec![vec![1, 3, 5, 7], vec![10, 11, 16, 20], vec![23, 30, 34, 60]];
        assert_eq!(true, Solution::search_matrix(matrix, 3));
    }

    #[test]
    fn case2() {
        let matrix = vec![vec![1]];
        assert_eq!(false, Solution::search_matrix(matrix, 0));
    }
}
